Clarify / add missing cloud management docs (#178)
dbos-inc/dbos-docs: docs/cloud-tutorials/application-management.md

To deploy your application to DBOS Cloud, run this command in its root directory:

```shell
npx dbos-cloud app deploy
```

Your application is deployed using the name in its `package.json`.
Application names should be between 3 and 30 characters and must contain only lowercase letters and numbers, dashes (`-`), and underscores (`_`). Application names are unique within an [organization](#users-and-organizations).

> When you deploy an application for the first time, the command prompts you to choose which [database instance](../cloud-tutorials/database-management.md) to connect your app to, or to provision one if you have none. Note this database instance (server) is different from the database name your application use (the `app_db_name` field in [`dbos-config.yaml`](../api-reference/configuration#database)).

Reviewer: ```suggestion
When you deploy an application for the first time, the command prompts you to choose which [database instance](../cloud-tutorials/database-management.md) to connect your app to, or to provision one if you have none. Note this database instance (server) is different from the database name your application uses (the `app_db_name` field in [`dbos-config.yaml`](../api-reference/configuration#database)).
```

> When you deploy an application for the first time, the command prompts you to choose which [database instance](../cloud-tutorials/database-management.md) to connect your app to, or to provision one if you have none. Note this database instance (server) is different from the database name your application uses (the `app_db_name` field in [`dbos-config.yaml`](../api-reference/configuration#database)).

Reviewer 2: Note that multiple applications can connect to the same database server, but must use separate databases within that server (app_db_name...)

:::tip
* You can specify a particular database instance through the `-d <database-instance-name>`.

> * During the first deployment, you can enable time travel for your application with `--enable-timetravel`. You can delete and re-deploy an existing application to enable time travel.

Reviewer: Just for my understanding: If I do not enable timetravel at first, it is not enough to just re-depoy with `--enable-timetravel`? I have to delete the application and re-deploy?

Author: yup you have to delete + redeploy

It then starts your application.
After your application is deployed, DBOS Cloud hosts it at this URL, which is also printed by the deploy command:

:::info

> Each application is allocated 1GB of disk space.

Reviewer 3: I'm seeing 500MB here: 
https://github.com/dbos-inc/dbos-cloud/blob/733e24f95ec258e06f12151428e42f3fa952c03b/services/applications/deploy.go#L22

Is the code lying?

Reviewer 4: I'm still hung up on the idea of "disk space".  You shouldn't be getting "disk space".  Is it that the application image should be less than 1GB (or 500MB)?

Author: > I'm seeing 500MB here: https://github.com/dbos-inc/dbos-cloud/blob/733e24f95ec258e06f12151428e42f3fa952c03b/services/applications/deploy.go#L22
> 
> Is the code lying?

No, you are right, but during the build step, we allow 1GB of disk to accommodate for dev dependencies. So a user could see >500MB from their app. But the deploy will fail if the archive size is more than 500MB. I will update now.*

> I'm still hung up on the idea of "disk space". You shouldn't be getting "disk space". Is it that the application image should be less than 1GB (or 500MB)?

I'll adjust based on Alex comment and say that `Each application can be up to 500MB in size`

Author: Also clarified this section

> Your application must not use more than 1GB of disk space.

Reviewer 5: Rephrase it to something like: "Each app is allocated with 1GB of disk space." so it sounds more positive.

> Your application must not consume more than 1GB of disk space.

Reviewer 6: Just curious - why is this a requirement? Wouldn't this detract some customers from using DBOS?

Reviewer 6: And isn't this requirement repeated below?

Author: Just our current configuration. Can evolve in the future.

Reviewer 4: Your application is allowed to consume "disk space"?!  What even is that?

Author: "must not use more .." updated

If you edit your application or schema, run `npx dbos-cloud app deploy` again to apply the latest migrations and upgrade to the latest version.

:::info
* You don't have to worry about changing database server connection parameters like `hostname` or `password` in [`dbos-config.yaml`](../api-reference/configuration.md) to deploy an application to the cloud&#8212;DBOS automatically applies the connection information of your cloud database instance.

> * You cannot change the database name (`app_db_name`) of a deployed application. You can delete and re-deploy the application.

Reviewer 2: ```suggestion
* You cannot change the application database (`app_db_name`) of a deployed application. You must delete and re-deploy the application.
```

> * Databases cannot be shared across applications.

Reviewer: This **might** need clarification/differentiation between database instance and app db name, similar to what you did above.

> * Databases (`app_db_name`) cannot be shared across applications.

Reviewer 2: ```suggestion
* While database instances can be shared across applications, application databases (`app_db_name`) cannot.
```

- To retrieve the status of a particular application, run [`npx dbos-cloud app status <app-name>`](../api-reference/cloud-cli.md#npx-dbos-cloud-app-status). To retrieve the statuses of all applications, run [`npx dbos-cloud app list`](../api-reference/cloud-cli.md#npx-dbos-cloud-app-list).

> Applications are run in microVMs sized with 1vCPU and 512MB of RAM. Note that files on disk are ephemeral and not persisted between deployments.

Reviewer 2: ```suggestion
Applications are run in [Firecracker microVMs](https://firecracker-microvm.github.io/) with 1vCPU and 512MB of RAM. Note that files on disk are ephemeral and not persisted between deployments.
```

You can redeploy a previous version of your application by passing `--previous-version <version-id>` to the [`app deploy`](../api-reference/cloud-cli.md#npx-dbos-cloud-app-deploy) command.

```shell
npx dbos-cloud app deploy --previous-version <version-id>
```

However, this will fail if the previous and current versions have different database schemas.

> Also, note DBOS Cloud will keep running Transact workflows from previous versions until they complete. This means you need to be mindful of non-backward compatible database schema migrations, as they could break running workflows.

Reviewer: Do we have any safeguards/warnings in place for that? Or would it just break a running workflow from an old version if I deploy a non-backward compatible schema migration while an old workflow is still running?

Reviewer 2: No safeguards at the moment. To be fair, this problem isn't unique to Transact--if you do rolling deployments of an app to any service, you'll hit the same problem with non-backwards-compatible migrations (which is why many shops won't do such migrations at all, they're too risky).

> Also, note DBOS Cloud will keep running Transact workflows from previous versions until they complete. This means you need to be mindful of non-backward compatible database schema migrations, as they could break running workflows.

Reviewer 2: ```suggestion
Also, note DBOS Cloud will keep running Transact workflows from previous versions until they complete. This means you need to be mindful of non-backward-compatible database schema migrations, as they could break running workflows.
```

For more information on schema management, see our [schema management guide](./database-management.md#database-schema-management).

### Deleting Applications

To delete an application, run:

```shell
npx dbos-cloud app delete <app-name>
```

> The application will be fully deleted once all its running workflows complete.

Reviewer 2: It's deleted immediately, workflows or no

Author: What I mean is that the application will be in state "DELETED", and a record still present & listed in `dbos-cloud app list`, until all workflows are completed

Reviewer 2: No, it won't, we changed that behavior (or if we didn't, we should), otherwise long-running workflows would block deletion forever

You can also delete the application database with the `--dropdb` argument:

```shell
npx dbos-cloud app delete <app-name> --dropdb
```

> This will delete the database and all associated DBOS roles.

Reviewer 2: ```suggestion
```
